smbios.c

```diff
@@ -0,0 +1,727 @@
+    context->ptr = NULL;
+    context->sversion = VALID_VERSION(version) ? SMBIOS_3_0 : version;
+
+    // we have a valid SMBIOS entry point?
+    #ifndef _WIN32
+    context->data = data + SMBIOS_HEADER_SIZE;
+    context->size = size - SMBIOS_HEADER_SIZE;
+
+    if (data[0] == '_' && data[1] == 'S' && data[2] == 'M' && data[3] == '_')
+    {
+        // version 2.x
+
+        // entry point length
+        if (data[5] != 0x1F)
+            return SMBERR_INVALID_DATA;
+        // entry point revision
+        if (data[10] != 0)
+            return SMBERR_INVALID_DATA;
+        // intermediate anchor string
+        if (data[16] != '_' || data[17] != 'D' || data[18] != 'M' || data[19] != 'I' || data[20] != '_')
+            return SMBERR_INVALID_DATA;
+
+        // get the SMBIOS version
+        context->oversion = data[6] << 8 | data[7];
+    }
+    else
+    if (data[0] == '_' && data[1] == 'S' && data[2] == 'M' && data[3] == '3' && data[4] == '_')
+    {
+        // version 3.x
+
+        // entry point length
+        if (data[6] != 0x18)
+            return SMBERR_INVALID_DATA;
+        // entry point revision
+        if (data[10] != 0x01)
+            return SMBERR_INVALID_DATA;
+
+        // get the SMBIOS version
+        context->oversion = data[7] << 8 | data[8];
+    }
+    else
+        return SMBERR_INVALID_DATA;
+    #else
+    RawSMBIOSData *smBiosData = nullptr;
+    smBiosData = (RawSMBIOSData *) data;
+
+    // get the SMBIOS version
+    context->oversion = smBiosData->SMBIOSMajorVersion << 8 | smBiosData->SMBIOSMinorVersion;
+    context->data = smBiosData->SMBIOSTableData;
+    context->size = smBiosData->Length;
+    #endif
+
+    if (!VALID_VERSION(context->oversion))
+        return SMBERR_INVALID_DATA;
+    if (context->sversion > context->oversion)
+        context->sversion = context->oversion;
+
+    return SMBERR_OK;
+}
+
+const char *smbios_get_string( const struct Entry *entry, int index )
+{
+    if (entry == NULL || index <= 0 || index > entry->string_count)
+        return NULL;
+
+    const char *ptr = entry->strings;
+    for (int i = 1; *ptr != 0 && i < index; ++i)
+    {
+        while (*ptr != 0) ++ptr;
+        ++ptr;
+    }
+    return ptr;
+}
+
+static const char *get_string( struct ParserContext *context, int index )
+{
+    return smbios_get_string(&context->entry, index);
+}
+
+int smbios_reset( struct ParserContext *context )
+{
+    if (context == NULL)
+        return SMBERR_INVALID_ARGUMENT;
+    if (context->data == NULL || context->failed)
+        return SMBERR_INVALID_DATA;
+    context->ptr = context->estart = context->eend = NULL;
+    return SMBERR_OK;
+}
+
+static uint8_t read_uint8(struct ParserContext *context)
+{
+    if (context->ptr + 1 >= context->eend)
+    {
+        context->failed = true;
+        return 0;
+    }
+    return *context->ptr++;
+}
+
+static uint16_t read_uint16(struct ParserContext *context)
+{
+    return read_uint8(context) | (read_uint8(context) << 8);
```

Comment:
I do wonder, do we potentially need to swap bytes on some systems, regarding big endian vs little endian?